acados/ocp_nlp/ocp_nlp_constraints_bgp.c

```diff
@@ -0,0 +1,1253 @@
+    }
+
+    if (nr > 0)
+    {
+        // ext_fun_type_in[0] = BLASFEO_DVEC;
+        // ext_fun_in[0] = memory->ux;  // ux: nu+nx
+
+        //
+        ext_fun_type_out[0] = IGNORE_ARGUMENT;
+        //
+        ext_fun_type_out[1] = BLASFEO_DMAT_ARGS;
+        struct blasfeo_dmat_args jac_r_tran_out;
+        jac_r_tran_out.A = &work->jacobian_quadratic;
+        jac_r_tran_out.ai = 0;
+        jac_r_tran_out.aj = 0;
+        ext_fun_out[1] = &jac_r_tran_out;  // jac': (nu+nx) * nr
+        model->nl_constr_r_fun_jac->evaluate(model->nl_constr_r_fun_jac, ext_fun_type_in, ext_fun_in, ext_fun_type_out, ext_fun_out);
+
+        // add SCQP Hessian contribution
+        for (int i = 0; i < nphi; i++) { 
+            double lam_i = blasfeo_dvecex1(memory->lam, 
+                    2 * (nb + ng) + nphi + i);
+
+            blasfeo_dgemm_nt(nv, nr, nr, lam_i, &work->jacobian_quadratic, 
+                    0, 0, &work->tmp_nr_nphi_nr, nr * i, 0, 0.0, 
+                    &work->tmp_nv_nr, 0, 0, &work->tmp_nv_nr, 0, 0);
+
+            blasfeo_dsyrk_ln(nv, nr, 1.0, &work->tmp_nv_nr, 0, 0,
+                    &work->jacobian_quadratic, 0, 0, 1.0, memory->RSQrq,
+                    0, 0, memory->RSQrq, 0, 0);
+        }
+    }
+
+    blasfeo_daxpy(nb + ng + nphi, -1.0, &work->tmp_ni, 0, &model->d, 0, 
+            &memory->fun, 0);
+    blasfeo_daxpy(nb + ng + nphi, -1.0, &model->d, nb + ng + nphi, 
+            &work->tmp_ni, 0, &memory->fun, nb + ng + nphi);
+
+    // soft
+    blasfeo_dvecad_sp(ns, -1.0, memory->ux, nu + nx, model->idxs, &memory->fun, 0);
+    blasfeo_dvecad_sp(ns, -1.0, memory->ux, nu + nx + ns, model->idxs, &memory->fun, nb + ng + nphi);
+
+    blasfeo_daxpy(2 * ns, -1.0, memory->ux, nu + nx, &model->d, 2 * nb + 2 * ng + 2 * nphi,
+                  &memory->fun, 2 * nb + 2 * ng + 2 * nphi);
+
+    // nlp_mem: ineq_adj
+    if (opts->compute_adj)
+    {
+        blasfeo_dvecse(nu + nx + 2 * ns, 0.0, &memory->adj, 0);
+        blasfeo_daxpy(nb+ng+nphi, -1.0, memory->lam, nb + ng + nphi, memory->lam, 0, &work->tmp_ni, 0);
+        blasfeo_dvecad_sp(nb, 1.0, &work->tmp_ni, 0, model->idxb, &memory->adj, 0);
+        blasfeo_dgemv_n(nu+nx, ng+nphi, 1.0, memory->DCt, 0, 0, &work->tmp_ni, nb, 1.0, &memory->adj,
+                        0, &memory->adj, 0);
+        // soft
+        blasfeo_dvecex_sp(ns, 1.0, model->idxs, memory->lam, 0, &memory->adj, nu + nx);
+        blasfeo_dvecex_sp(ns, 1.0, model->idxs, memory->lam, nb+ng+nphi, &memory->adj, nu+nx+ns);
+        blasfeo_daxpy(2 * ns, 1.0, memory->lam, 2 * nb + 2 * ng + 2 * nphi, &memory->adj, nu + nx,
+                      &memory->adj, nu + nx);
+    }
+
+    if (opts->compute_hess)
+    {
+        printf("\nerror: compute_hess!=0 not supported (yet) in ocp_nlp_constraints_bgp\n");
+        exit(1);
+    }
+
+    return;
+}
+
+
+
+void ocp_nlp_constraints_bgp_compute_fun(void *config_, void *dims_, void *model_, void *opts_, void *memory_, void *work_)
```

Comment:
Could you implement this, since how this constraint module is fixed?

This is just computing the value of the constraint function. It may be good to add a new external function type to return just the function (no jac or hess), one for `phi` and one for `r`.

Reply:
I gave it a shot in 95338d58be, but I don't know how to test it :P Why do we need a function for r too?

Reply:
Since the nonlinear function is $\phi(r(x,u))$, I (apparently wrongly) assumed that $\phi$ was defined accordingly also in the C code, i.e. $\phi: R^{nr} \Rightarrow R^{nphi}$. But in your code it looks like it is defined as the composite function $\phi \circ r: R^{nx+nu} \Rightarrow R^{nphi}$.

Can you clarify on this point?

Reply:
`nl_constr_phi_jac` takes as input $x$, $u$, $z$ and $p$ and outputs $\phi(r(x,u,z))$, $\frac{d \phi(r(x,u,z))}{d x,u}$, $\frac{d \phi(r(x,u,z))}{d z}$ and the Hessians $\frac{d^2 \phi_i}{d r}$, which together with $\frac{d r(x,u,z)}{d x,u}$ are the quantity necessary to build the constraint linearization and the SCQP Hessian contribution.

Reply:
Thanks for the clarification. I think it is a bit confusing, as there is in my opinion a clash between $\phi$ and the composite function $\phi \circ r$. Also the way this is implemented made me believe something wrong at fist sight (shame on me for not digging deeper in the code before). I will explain in a new thread below.

Reply:
I thought mathematically speaking this is consistent since $\phi$ only denotes the image of the function (which is the same for $\phi(r(x,u,z))$ and $\phi(r)$), but we can make it more explicit/clear if you like.

Regarding the whole approach, this should be following Robin's notation in the most general form I could imagine :P